### PR TITLE
machines: Fix VM deletion bugs

### PR DESCRIPTION
**Pull request by skobyda**

### CODE REVIEW

File: pkg/machines/libvirt-dbus.js [hunk 1]
```diff
@@ -362,6 +362,12 @@ const LIBVIRT_DBUS_PROVIDER = {
                     storageVolPromises.push(
                         call(connectionName, '/org/libvirt/QEMU', 'org.libvirt.Connect', 'StorageVolLookupByPath', [disk.source.file], { timeout, type: 's' })
                                 .then(volPath => call(connectionName, volPath[0], 'org.libvirt.StorageVol', 'Delete', [0], { timeout, type: 'u' }))
+                                .catch(ex => {
+                                    if (!ex.message.includes("no storage vol with matching path"))
+                                        return Promise.reject(ex);
+                                    else
+                                        return cockpit.file(disk.source.file, { superuser: "try" }).replace(null); // delete key file
```
Comment (martinpitt) [inline pkg/machines/libvirt-dbus.js:369]: This looked a bit brittle at first, but I confirmed that trying to delete a non-existing file succeeds.

**Review by martinpitt — CHANGES_REQUESTED** [review]
Thanks!

File: pkg/machines/libvirt-dbus.js [hunk 2]
```diff
@@ -362,6 +362,12 @@ const LIBVIRT_DBUS_PROVIDER = {
                     storageVolPromises.push(
                         call(connectionName, '/org/libvirt/QEMU', 'org.libvirt.Connect', 'StorageVolLookupByPath', [disk.source.file], { timeout, type: 's' })
                                 .then(volPath => call(connectionName, volPath[0], 'org.libvirt.StorageVol', 'Delete', [0], { timeout, type: 'u' }))
+                                .catch(ex => {
+                                    if (!ex.message.includes("no storage vol with matching path"))
```
[the remark this answers is not in the document]
Comment (skobyda) [reply pkg/machines/libvirt-dbus.js:366]: Fixed

File: test/verify/check-machines [hunk 3]
```diff
@@ -1396,6 +1403,28 @@ class TestMachines(MachineCase, StorageHelpers, NetworkHelpers):
         self.waitVmRow(name, 'system', False)
         b.wait_not_present('#vm-{0}-state button:contains("view more")'.format(name))
 
+        # Try to delete VM which has a disk defined with a file as a source instead of a storage volume
+        # In this case, the libvirt's storage volume deletion API would fall but cockpit would fall back to deleting the file directly
```
Comment (martinpitt) [inline test/verify/check-machines:1407]: "would fall" → "would fail"

File: test/verify/check-machines [hunk 4]
```diff
@@ -1396,6 +1403,28 @@ class TestMachines(MachineCase, StorageHelpers, NetworkHelpers):
...
+        m.execute("virsh attach-disk --domain {0} --source {1} --target vdb --persistent".format(name, disk))
+
+        b.reload()
+        self.browser.enter_page('/machines')
+        self.goToMainPage()
+        b.click("#vm-{0}-action-kebab button".format(name))
+        b.click("#vm-{0}-delete".format(name))
+
+        b.wait_visible("#vm-{0}-delete-modal-dialog".format(name))
+        self.assertTrue(lambda: checkFilePresent(disk))
```
Comment (martinpitt) [inline test/verify/check-machines:1422]: simpler: `m.execute("test -e " + disk)`

File: test/verify/check-machines [hunk 5]
```diff
@@ -1396,6 +1403,28 @@ class TestMachines(MachineCase, StorageHelpers, NetworkHelpers):
...
+        b.click("#vm-{0}-delete-modal-dialog button:contains(Delete)".format(name))
+        b.wait_not_present("#vm-{0}-delete-modal-dialog".format(name))
+        self.assertTrue(lambda: not checkFilePresent(disk))
```
Comment (martinpitt) [inline test/verify/check-machines:1425]: This is a bit weird -- please use `assertFalse` to check for negative conditions.  But moreover, `m.execute("test ! -e " + disk`

**Review by martinpitt — CHANGES_REQUESTED** [review]
Thanks! Looks good in general (I have a few nitpicks), but TestMachines.testDelete fails a lot, this needs fixing.

**Review by martinpitt — CHANGES_REQUESTED** [review]
Code changes look good, thank you! But there's still at least one 
[testDelete failure](https://logs.cockpit-project.org/logs/pull-15102-20210126-123711-ca920132-rhel-8-4/log.html#314) (tests are still running everywhere, though)

**Review by KKoukiou — CHANGES_REQUESTED** [review]
Simon the test is already there, no need for a new one.
Just remove the set_checked after:
        
        # HACK: Don't delete disks here because they will fail with 'storage volume does not exist'
        # This can be reverted when https://github.com/cockpit-project/cockpit/pull/15102 gets merged

**Review by KKoukiou — APPROVED** [review]
It looks good now! thanks for fixing this :)

**Pull request merged**